### PR TITLE
Chapter Narration Page

### PR DESCRIPTION
**Pull request by mXaln**
This change is [Reviewable](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723)

### CODE REVIEW

File: jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/waveform/ContinuousWaveformLayer.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package org.wycliffeassociates.otter.jvm.controls.waveform
+
+import javafx.scene.canvas.Canvas
+import javafx.scene.canvas.GraphicsContext
+import javafx.scene.paint.Paint
+import org.wycliffeassociates.otter.common.recorder.ActiveRecordingRenderer
+
+private const val USHORT_SIZE = 65535.0
```
Comment (AnonymousWalker) [inline jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/waveform/ContinuousWaveformLayer.kt]: maybe ` UShort.MAX_VALUE ` instead ?

**Review by mXaln — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723)* status: 0 of 86 files reviewed, 2 unresolved discussions (waiting on @AnonymousWalker)

___
*[`common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/InsertTakeAudio.kt` line 25 at r2](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723#gh-1158603291:-NSoc-4E2Wqntjj7cz8N:bqh01mt) ([raw file](https://github.com/Bible-Translation-Tools/Orature/blob/abbf5138295b71b7d2b49198a8e07a789fc68fc6/common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/InsertTakeAudio.kt#L25)):*
<details><summary><i>Previously, AnonymousWalker (Tony Tran) wrote…</i></summary><blockquote>

do we need to filter out the deleted chunks?
</blockquote></details>

What do you mean by deleted chunks?
___
*[`jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/waveform/ContinuousWaveformLayer.kt` line 8 at r2](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723#gh-1158668982:-NSoc2Rz5wEgAtywuHlk:bcx7it9) ([raw file](https://github.com/Bible-Translation-Tools/Orature/blob/abbf5138295b71b7d2b49198a8e07a789fc68fc6/jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/waveform/ContinuousWaveformLayer.kt#L8)):*
<details><summary><i>Previously, AnonymousWalker (Tony Tran) wrote…</i></summary><blockquote>

maybe ` UShort.MAX_VALUE ` instead ?
</blockquote></details>

Nice shortcut!

**Review by AnonymousWalker — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723)* status: 0 of 86 files reviewed, 1 unresolved discussion

___
*[`common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/InsertTakeAudio.kt` line 25 at r2](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723#gh-1158603291:-NSq0d6JC9tIMeEA4zTl:b-agbdo6) ([raw file](https://github.com/Bible-Translation-Tools/Orature/blob/abbf5138295b71b7d2b49198a8e07a789fc68fc6/common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/InsertTakeAudio.kt#L25)):*
<details><summary><i>Previously, mXaln (Maxim) wrote…</i></summary><blockquote>

What do you mean by deleted chunks?
</blockquote></details>

oh it seems like narration doesn't have chunks and we won't have the "reset" button for verse/chunk. My bad

File: common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/PcmTakeTransformer.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+     */
+    @Throws(IllegalArgumentException::class)
+    fun postTransform(take: Take): Take {
+        return originalTake?.let { origTake ->
+            val format = AudioFileFormat.of(origTake.file.extension)
+            if (format == AudioFileFormat.PCM) {
+                audioConverter.wavToPcm(take.file, origTake.file).blockingGet()
+            }
+            origTake
+        } ?: throw IllegalArgumentException("postTransform is called before preTransform")
```
Comment (AnonymousWalker) [inline common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/PcmTakeTransformer.kt:68]: do we need to reset `originalTake = null` after finishing postTranform()?

File: common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/SplitAudioOnCues.kt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    private fun writeAudio(source: AudioFile, target: AudioFile, startEnd: Pair<Int, Int>) {
+        val sourceReader = source.reader(startEnd.first, startEnd.second)
+        val targetWriter = target.writer()
+
+        sourceReader.use { reader ->
+            reader.open()
+            targetWriter.use { writer ->
+                val buffer = ByteArray(10240)
```
Comment (AnonymousWalker) [inline common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/SplitAudioOnCues.kt]: you may want a `const val BUFFER_SIZE` or a comment to explain why we chose 10240 - magic number

File: common/src/main/kotlin/org/wycliffeassociates/otter/common/recorder/ActiveRecordingRenderer.kt [hunk 4]
```diff
@@ -92,13 +109,18 @@ class ActiveRecordingRenderer(
...
+            .also(compositeDisposable::add)
     }
 
     /** Clears rendered data from buffer */
     fun clearData() {
         floatBuffer.clear()
+        audioData.clear()
+    }
+
+    fun disposeOfListeners() {
```
Comment (AnonymousWalker) [inline common/src/main/kotlin/org/wycliffeassociates/otter/common/recorder/ActiveRecordingRenderer.kt]: disposeListeners ?

File: jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/NarrationRecordListView.kt [hunk 5]
```diff
@@ -18,20 +18,25 @@
...
 class NarrationRecordListView<T>(items: ObservableList<T>? = null) : ListView<T>(items) {
+    val isRecordingProperty = SimpleBooleanProperty()
+
     init {
         addClass("wa-list-view")
         orientation = Orientation.HORIZONTAL
+
+        disableProperty().bind(isRecordingProperty)
+        styleProperty().bind(isRecordingProperty.stringBinding {
+            if (it == true) "-fx-opacity: 1" else ""
```
Comment (AnonymousWalker) [inline jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/NarrationRecordListView.kt:38]: using a style class like how you've done in other places would be cleaner

File: jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/NarrationTextListView.kt [hunk 6]
```diff
@@ -41,10 +42,12 @@ class NarrationTextListView<T>(items: ObservableList<T>? = null) : ListView<T>(i
     init {
         addClass("wa-list-view")
 
-        initialSelectedItemProperty.onChange {
-            cardIsOutOfViewProperty.set(false)
-            selectionModel.select(it)
-            scrollTo(it)
+        initialSelectedItemProperty.onChangeAndDoNow {
```
Comment (AnonymousWalker) [inline jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/NarrationTextListView.kt]: do we need onChangeAndDoNowWithDispose ?

File: jvm/controls/src/main/resources/css/narration.css [hunk 7]
```diff
@@ -172,51 +175,88 @@
...
+    -fx-background-color: #001533;
+    -fx-alignment: bottom-center;
+}
+
+.narration-record__volumebar-filler {
+    -fx-pref-width: 16px;
+    -fx-background-color: #001533;
+}
+
+.narration-list__buttons .btn.recording {
```
Comment (AnonymousWalker) [inline jvm/controls/src/main/resources/css/narration.css]: sounds like we should have a class name "recording-btn" instead

File: jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/components/NarrationTextCell.kt [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    override fun updateItem(item: ChunkData?, empty: Boolean) {
+        super.updateItem(item, empty)
+
+        if (empty || item == null) {
+            graphic = null
+            return
+        }
+
+        view.isSelectedProperty.set(isSelected)
+        view.isLastVerseProperty.set(index == listView.items.lastIndex)
```
Comment (AnonymousWalker) [inline jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/components/NarrationTextCell.kt]: could we move this inside the `apply { }` block below?

**Review by AnonymousWalker — CHANGES_REQUESTED** [review]

File: common/src/main/kotlin/org/wycliffeassociates/otter/common/recorder/ActiveRecordingRenderer.kt (hunk 4 above)
Comment (jsarabia) [reply common/src/main/kotlin/org/wycliffeassociates/otter/common/recorder/ActiveRecordingRenderer.kt]: @AnonymousWalker disposeListeners doesn't sound natural, if we were to keep the word dispose, I think disposeOf is appropriate.

an alternative could be releaseListeners, closeListeners

**Review by jsarabia — COMMENTED** [review]
> Reviewed 2 of 72 files at r1, 1 of 69 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723)* status: 3 of 90 files reviewed, 9 unresolved discussions (waiting on @AnonymousWalker and @mXaln)

___
*[`common/audio/src/main/kotlin/org/wycliffeassociates/otter/common/audio/wav/WavFileReader.kt` line 37 at r4](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723#-NTOW0X-4qwtby-7hvx7:-NTOW0X-4qwtby-7hvx8:b-gdk2xd) ([raw file](https://github.com/Bible-Translation-Tools/Orature/blob/2d5dc63378879342fecbc5f93dbda42494e712ca/common/audio/src/main/kotlin/org/wycliffeassociates/otter/common/audio/wav/WavFileReader.kt#L37)):*
> ```Kotlin
> 
>     override val totalFrames: Int
>         get() = end - start
> ```

Is this changed because it is a bug? Can you describe why wav.totalFrames is erroneous? 

I don't see any reason for the changes to this file for the sake of Narration. I understand the principle to clean up code as we go, however, it is important to treat bugs as separate PRs. They get attached to a commit which can be identified, reverted, provides a message that hopefully explains the bug and why the change, etc.

As a reviewer, I can't see why this change was made. The change here specifically is not equivalent to the code before. If it was a bug, I'd like to know a little more about why and how it is demonstrated to be incorrect. If it's a change because it sounds reasonable or looks good- how do I know that the change does not introduce a bug?

**Review by mXaln — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723)* status: 3 of 90 files reviewed, 9 unresolved discussions (waiting on @AnonymousWalker and @jsarabia)

___
*[`common/audio/src/main/kotlin/org/wycliffeassociates/otter/common/audio/wav/WavFileReader.kt` line 37 at r4](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723#-NTOW0X-4qwtby-7hvx7:-NTP434mBwNsd2I5KKR6:b-7ijmx) ([raw file](https://github.com/Bible-Translation-Tools/Orature/blob/2d5dc63378879342fecbc5f93dbda42494e712ca/common/audio/src/main/kotlin/org/wycliffeassociates/otter/common/audio/wav/WavFileReader.kt#L37)):*
<details><summary><i>Previously, jsarabia (Joe) wrote…</i></summary><blockquote>

Is this changed because it is a bug? Can you describe why wav.totalFrames is erroneous? 

I don't see any reason for the changes to this file for the sake of Narration. I understand the principle to clean up code as we go, however, it is important to treat bugs as separate PRs. They get attached to a commit which can be identified, reverted, provides a message that hopefully explains the bug and why the change, etc.

As a reviewer, I can't see why this change was made. The change here specifically is not equivalent to the code before. If it was a bug, I'd like to know a little more about why and how it is demonstrated to be incorrect. If it's a change because it sounds reasonable or looks good- how do I know that the change does not introduce a bug?
</blockquote></details>

I made this change, because when I load a section of a file, I want totalFrames to reflect the size of loaded audio. I needed this in audio player. When I load a section of a file, I need to know totalFrames of that section to properly play the audio.
___
*[`common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/PcmTakeTransformer.kt` line 68 at r3](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723#gh-1164568205:-NTP5BK2EuxCl3A7Yh-t:b-ifijkq) ([raw file](https://github.com/Bible-Translation-Tools/Orature/blob/4ad183a4c6eecf47b3bd451503fd21756dcf1e69/common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/PcmTakeTransformer.kt#L68)):*
<details><summary><i>Previously, AnonymousWalker (Tony Tran) wrote…</i></summary><blockquote>

do we need to reset `originalTake = null` after finishing postTranform()?
</blockquote></details>

What is the benefit of making it null?
___
*[`common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/SplitAudioOnCues.kt` line 68 at r3](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723#gh-1164570379:-NTP5pvNCGJZLI51rqh_:b9kudgi) ([raw file](https://github.com/Bible-Translation-Tools/Orature/blob/4ad183a4c6eecf47b3bd451503fd21756dcf1e69/common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/SplitAudioOnCues.kt#L68)):*
<details><summary><i>Previously, AnonymousWalker (Tony Tran) wrote…</i></summary><blockquote>

you may want a `const val BUFFER_SIZE` or a comment to explain why we chose 10240 - magic number
</blockquote></details>

Makes sense. Done
___
*[`common/src/main/kotlin/org/wycliffeassociates/otter/common/recorder/ActiveRecordingRenderer.kt` line 121 at r3](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723#gh-1164572251:-NTP68zcExLzVAzYMYh7:b-896fix) ([raw file](https://github.com/Bible-Translation-Tools/Orature/blob/4ad183a4c6eecf47b3bd451503fd21756dcf1e69/common/src/main/kotlin/org/wycliffeassociates/otter/common/recorder/ActiveRecordingRenderer.kt#L121)):*
<details><summary><i>Previously, jsarabia (Joe) wrote…</i></summary><blockquote>

@AnonymousWalker disposeListeners doesn't sound natural, if we were to keep the word dispose, I think disposeOf is appropriate.

an alternative could be releaseListeners, closeListeners
</blockquote></details>

Done.
___
*[`jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/NarrationRecordListView.kt` line 38 at r3](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723#gh-1164575174:-NTP7bBt8SCxD_4IQtYz:b-2x74mv) ([raw file](https://github.com/Bible-Translation-Tools/Orature/blob/4ad183a4c6eecf47b3bd451503fd21756dcf1e69/jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/NarrationRecordListView.kt#L38)):*
<details><summary><i>Previously, AnonymousWalker (Tony Tran) wrote…</i></summary><blockquote>

using a style class like how you've done in other places would be cleaner
</blockquote></details>

I'm using bind to styleProperty, because binding is using weak listener that can be garbage collected. When onChange is not
___
*[`jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/NarrationTextListView.kt` line 45 at r3](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723#gh-1164576784:-NTPE76W1_-8XCP_mnaO:b-896fix) ([raw file](https://github.com/Bible-Translation-Tools/Orature/blob/4ad183a4c6eecf47b3bd451503fd21756dcf1e69/jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/NarrationTextListView.kt#L45)):*
<details><summary><i>Previously, AnonymousWalker (Tony Tran) wrote…</i></summary><blockquote>

do we need onChangeAndDoNowWithDispose ?
</blockquote></details>

Done.
___
*[`jvm/controls/src/main/resources/css/narration.css` line 260 at r3](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723#gh-1164580859:-NTPFDdPD_5QdHE02Qw0:b-896fix) ([raw file](https://github.com/Bible-Translation-Tools/Orature/blob/4ad183a4c6eecf47b3bd451503fd21756dcf1e69/jvm/controls/src/main/resources/css/narration.css#L260)):*
<details><summary><i>Previously, AnonymousWalker (Tony Tran) wrote…</i></summary><blockquote>

sounds like we should have a class name "recording-btn" instead
</blockquote></details>

Done.
___
*[`jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/components/NarrationTextCell.kt` line 43 at r3](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723#gh-1164582457:-NTPFF1bEMi80Y_ECjgZ:b-ukgfbb) ([raw file](https://github.com/Bible-Translation-Tools/Orature/blob/4ad183a4c6eecf47b3bd451503fd21756dcf1e69/jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/components/NarrationTextCell.kt#L43)):*
<details><summary><i>Previously, AnonymousWalker (Tony Tran) wrote…</i></summary><blockquote>

could we move this inside the `apply { }` block below?
</blockquote></details>

Probably this is out of sync?

**Review by jsarabia — COMMENTED** [review]
> Reviewed 4 of 72 files at r1, 3 of 69 files at r2, 3 of 13 files at r3, 2 of 10 files at r4.
> *[Reviewable](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723)* status: 13 of 90 files reviewed, 10 unresolved discussions (waiting on @AnonymousWalker and @mXaln)

___
*[`jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/CanvasFragment.kt` line 30 at r4](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723#-NTQEx0x1wG8l-bF7789:-NTQEx0x1wG8l-bF778A:b-yt58uo) ([raw file](https://github.com/Bible-Translation-Tools/Orature/blob/2d5dc63378879342fecbc5f93dbda42494e712ca/jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/CanvasFragment.kt#L30)):*
> ```Kotlin
> import tornadofx.*
> 
> class CanvasFragment : StackPane() {
> ```

If we're going to hardcode a waveform classname in this, might as well call this a ActiveRecordingWaveformPane or something like that

**Review by mXaln — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723)* status: 13 of 90 files reviewed, 10 unresolved discussions (waiting on @AnonymousWalker and @jsarabia)

___
*[`jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/CanvasFragment.kt` line 30 at r4](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/723#-NTQEx0x1wG8l-bF7789:-NTSoATv6dk_kHXvQ_ww:bltw0v1) ([raw file](https://github.com/Bible-Translation-Tools/Orature/blob/2d5dc63378879342fecbc5f93dbda42494e712ca/jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/CanvasFragment.kt#L30)):*
<details><summary><i>Previously, jsarabia (Joe) wrote…</i></summary><blockquote>

If we're going to hardcode a waveform classname in this, might as well call this a ActiveRecordingWaveformPane or something like that
</blockquote></details>

But this canvas fragment is not always a waveform. It can be a volume bar. This is a basic class to draw on canvas

**Pull request closed without merging**